symengine/series_generic.cpp

```diff
@@ -2,155 +2,242 @@
-        if (first) {
-            if (it.second < 0)
-                o << "-";
-        }
-        else {
-            if (it.second < 0)
-                o << " - ";
-            else
-                o << " + ";
-        }
-        first = false;
-        if (it.first == 0) {
-            o << mp_abs(it.second);
-            continue;
+RCP<const UnivariatePolynomial> UnivariateSeries::trunc_poly(const RCP<const Symbol> &var, const map_int_Expr &d, unsigned pr) {
+    map_int_Expr dict_trunc;
+    unsigned int max = 0;
+    for (const auto &it : d) {
+        if ((unsigned)it.first < pr) {
+            if (max < (unsigned)it.first)
+                max = (unsigned)it.first;
+            dict_trunc[it.first] = it.second;
         }
-        if (mp_abs(it.second) == 1)
-            o << var_->get_name();
-        else
-            o << mp_abs(it.second) << "*" << var_->get_name();
-        if (it.first > 1)
-            o << "**" << it.first;
     }
-    if (o.str() != "0")
-        o << " + O(" << var_->get_name() << "**" << prec_ << ")";
-    return o.str();
-}
-
-RCP<const UnivariateSeries> add_uni_series (const UnivariateSeries& a, const UnivariateSeries& b)
-{
-    map_uint_mpz dict;
-    SYMENGINE_ASSERT(a.var_->get_name() == b.var_->get_name())
-    unsigned int minprec = (a.prec_ < b.prec_)? a.prec_ : b.prec_;
-    for (const auto &it : a.poly_->get_dict()) {
-        if (it.first >= minprec)
-            break;
-        dict[it.first] = it.second;
+    return univariate_polynomial(var, max, std::move(dict_trunc));
+}
+
+unsigned UnivariateSeries::ldegree(const UnivariateExprPolynomial &s) {
+    return s.get_univariate_poly()->get_dict().begin()->first;
+}
+
+UnivariateExprPolynomial UnivariateSeries::mul(const UnivariateExprPolynomial &a, const UnivariateExprPolynomial &b, unsigned prec) {
+    unsigned int exp;
+    map_int_Expr p;
+
+    for (auto &it1 : a.get_univariate_poly()->get_dict()) {
+        for (auto &it2 : b.get_univariate_poly()->get_dict()) {
+            exp = it1.first + it2.first;
+            if (exp < prec) {
+                p[exp] = it1.second * it2.second;
+            } else {
+                break;
+            }
+        }
     }
+    return UnivariateExprPolynomial(UnivariatePolynomial::from_dict(a.get_univariate_poly()->get_var(), std::move(p)));
+}
 
-    unsigned int max = 0;
-    for (const auto &it : b.poly_->get_dict()) {
-        if (it.first >= minprec)
-            break;
-        dict[it.first] += it.second;
-        if (dict[it.first] != 0 and it.first > max)
-            max = it.first;
+UnivariateExprPolynomial UnivariateSeries::pow(const UnivariateExprPolynomial &s, int n, unsigned prec) {
+    if (is_a<const Constant>(*s.get_basic()))
+        throw std::runtime_error("Polynomial cannot be a constant");
+    if (s.get_univariate_poly()->get_var() != symbol("x"))
+        throw std::runtime_error("Polynomial is not x");
+    if (n < 0)
+        throw std::runtime_error("Not Implemented");
```

Comment:
If `n` is negative, you have to check whether `s` is a `var` and handle that case